src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1839 @@
+            // be in another vreg's spillslot at the block beginning.
+            for vreg in self.func.branch_blockparams(block, inst, succ_idx).iter() {
+                if self.vreg_spillslots[vreg.vreg()].is_invalid() {
+                    self.vreg_spillslots[vreg.vreg()] = self.allocstack(vreg);
+                    trace!("Block arg {:?} is going to be in {:?}", vreg, Allocation::stack(self.vreg_spillslots[vreg.vreg()]));
+                }
+                if self.temp_spillslots[vreg.class()].len() == next_temp_idx[vreg.class()] {
+                    let newslot = self.allocstack(vreg);
+                    self.temp_spillslots[vreg.class()].push(newslot);
+                }
+                let temp_slot = self.temp_spillslots[vreg.class()][next_temp_idx[vreg.class()]];
+                let temp = Allocation::stack(temp_slot);
+                next_temp_idx[vreg.class()] += 1;
+                let vreg_spill = Allocation::stack(self.vreg_spillslots[vreg.vreg()]);
+                trace!("{:?} which is going to be in {:?} inserting move to {:?}", vreg, vreg_spill, temp);
+                self.add_move_later(inst, vreg_spill, temp, vreg.class(), InstPosition::Before, false);
+            }
+        }
+    
+        reset_temp_idx(&mut next_temp_idx);
+
+        for (succ_idx, succ) in self.func.block_succs(block).iter().enumerate() {
+            let succ_params = self.func.block_params(*succ);
+
+            // Move from temporaries to block param spillslots.
+            for (pos, vreg) in self.func.branch_blockparams(block, inst, succ_idx).iter().enumerate() {
+                let succ_param_vreg = succ_params[pos];
+                let param_alloc = Allocation::stack(self.vreg_spillslots[succ_param_vreg.vreg()]);
+                let temp_slot = self.temp_spillslots[vreg.class()][next_temp_idx[vreg.class()]];
+                let temp = Allocation::stack(temp_slot);
+                self.vreg_allocs[vreg.vreg()] = temp;
+                next_temp_idx[vreg.class()] += 1;
+                trace!(" --- Placing branch arg {:?} in {:?}", vreg, temp);
+                trace!("{:?} which is now in {:?} inserting move to {:?}", vreg, temp, param_alloc);
+                self.add_move_later(inst, temp, param_alloc, vreg.class(), InstPosition::Before, false);
+            }
+        }
+
+        reset_temp_idx(&mut next_temp_idx);
+
+        for (succ_idx, _) in self.func.block_succs(block).iter().enumerate() {
+            for vreg in self.func.branch_blockparams(block, inst, succ_idx).iter() {
+                // All branch arguments should be in their spillslots at the end of the function.
+                self.vreg_allocs[vreg.vreg()] = Allocation::stack(self.vreg_spillslots[vreg.vreg()]);
+                self.live_vregs.insert(*vreg);
+            }
+        }
+    }
+
+    fn alloc_inst(&mut self, block: Block, inst: Inst) -> Result<(), RegAllocError> {
```

Comment:
The order of operations here seems to be incorrect since it could fail to allocate an instruction with an early-fixed-def if that register was selected by a late-nonfixed-def that was previously processed.

You can take advantage of the fact that early defs and late defs are not allowed to overlap. You can think of it as the live range of a def always extends to the end of the instruction, no matter whether it is early or late. The same applies to late uses: the vreg comes from a previous instruction so it is "in use" in both the early and late part of the instruction.

A better ordering of operations would be to separately track which registers are available in the early and late part of the instruction.
- First process all (early & late, use & def) fixed constraints & clobbers. This removes the fixed registers from the set of available registers. For early use it only removes registers from the early set. For late def it only removes registers from the late set. For early def & late use it removes from both.
- Then process all remaining defs & uses, assigning from the remaining available registers.

Reply:
> The order of operations here seems to be incorrect since it could fail to allocate an instruction with an early-fixed-def if that register was selected by a late-nonfixed-def that was previously processed.

The algorithm allocates late defs first, then deallocates them.
All late defs are freed before the early fixed defs are processed, so this isn't possible.
I don't think order is incorrect.

Reply:
Remember that early defs are live until the end of the instruction (the def starts at the early point, but also cover the late point). So it's invalid to assign the same register to an early def and a late def because the live ranges overlap.

Reply:
Oh, I thought an early def that isn't used simply doesn't have its liverange extended anywhere, allowing its allocation to be free for a late operand.
Okay, that makes sense.